exist-core/src/main/java/org/exist/http/RESTServer.java

```diff
@@ -1162,61 +1191,66 @@ public void doDelete(final DBBroker broker, final Txn transaction, final String
-                    break; // found a binary file with mime-type xquery or XML file with mime-type xproc
-                } else if (null != resource) {
-                    // not an xquery or xproc resource. This means we have a path
-                    // that cannot contain an xquery or xproc object even if we keep
-                    // moving up the path, so bail out now
-                    lockedDocument.close();
-                    lockedDocument = null;
-                    resource = null;
-                    break;
-                }
-                servletPath = servletPath.removeLastSegment();
-                if (servletPath == XmldbURI.EMPTY_URI) {
-                    break;
-                }
+        XmldbURI servletPath = path;
+        LockedDocument lockedDocument = null;
+        DocumentImpl resource = null;
+        // work up the url path to find an
+        // xquery resource
+        while (resource == null) {
+            // traverse up the path looking for xquery objects
+
+            lockedDocument = broker.getXMLResource(servletPath, LockMode.READ_LOCK);
+            resource = lockedDocument == null ? null : lockedDocument.getDocument();
+            if (resource != null
+                    && (resource.getResourceType() == DocumentImpl.BINARY_FILE
+                    && xqueryType.equals(resource.getMimeType()))) {
+                break; // found a binary file with mime-type xquery or XML file with mime-type xproc
+            } else if (resource != null) {
+                // not an xquery or xproc resource. This means we have a path
+                // that cannot contain an xquery or xproc object even if we keep
+                // moving up the path, so bail out now
+                lockedDocument.close();
+                lockedDocument = null;
+                resource = null;
+                break;
             }
-            // xquery binary file found
-            if (resource != null) {
-                // found an XQuery resource, fixup request values
-                final String pathInfo = path.trimFromBeginning(servletPath).toString();
-                final Properties outputProperties = new Properties(defaultOutputKeysProperties);
-                try {
-                    // Execute the XQuery
-                    executeXQuery(broker, transaction, resource, request, response,
-                            outputProperties, servletPath.toString(), pathInfo);
-                } catch (final XPathException e) {
-                    writeXPathExceptionHtml(response, HttpServletResponse.SC_BAD_REQUEST, "UTF-8", null, path.toString(), e);
-                } finally {
-                    lockedDocument.close();
-                }
-                return true;
+            servletPath = servletPath.removeLastSegment();
+            if (servletPath == XmldbURI.EMPTY_URI) {
+                break;
             }
         }
-        return false;
+
+        // no xquery binary file found
```

Comment:
This should be moved up I think to line 1217.

Reply:
This is a safeguard against the while loop breaking while `resource` still being null.